src/textual/document/_wrapped_document.py

```diff
@@ -0,0 +1,154 @@
+
+    @property
+    def lines(self) -> list[list[str]]:
+        """The lines of the wrapped version of the Document.
+
+        Each index in the returned list represents a line index in the raw
+        document. The list[str] at each index is the content of the raw document line
+        split into multiple lines via wrapping.
+        """
+        wrapped_lines = []
+        append = wrapped_lines.append
+        for line_index, line in enumerate(self.document.lines):
+            divided = Text(line).divide(self._wrap_offsets[line_index])
+            append([section.plain for section in divided])
+        return wrapped_lines
+
+    def refresh_range(
+        self,
+        start: Location,
+        old_end: Location,
+        new_end: Location,
+    ) -> None:
+        """Incrementally recompute wrapping based on a performed edit.
+
+        This must be called *after* the source document has been edited.
+
+        Args:
+            start: The start location of the edit that was performed in document-space.
+            old_end: The old end location of the edit in document-space.
+            new_end: The new end location of the edit in document-space.
+        """
+
+        # Get all the text on the lines between start and end in document space
+        start_row, _ = start
+        end_row, _ = new_end
+
+        # +1 since we go to the start of the next row, and +1 for inclusive.
+        new_lines = self.document.lines[start_row : end_row + 2]
+
+        new_wrap_offsets = []
+        append_wrap_offset = new_wrap_offsets.append
+        for line_index, line in enumerate(new_lines, start_row):
+            wrap_offsets = divide_line(line, self._width)
+            append_wrap_offset(wrap_offsets)
+
+        # Replace the range start->old with the new wrapped lines
+        old_end_row, _ = old_end
+        self._wrap_offsets[start_row:old_end_row] = new_wrap_offsets
+
+    def offset_to_line_index(self, offset: int) -> int:
+        """Given an offset within the wrapped/visual display of the document,
+        return the corresponding line index.
+
+        Args:
+            offset: The y-offset within the document.
+
+        Raises:
+            ValueError: When the given offset does not correspond to a line
+                in the document.
+
+        Returns:
+            The line index corresponding to the given y-offset.
+        """
+
+        def invalid_offset_error():
+            raise ValueError(
+                f"No line exists at wrapped document offset {offset!r}. "
+                f"Document wrapped with width {self._width!r}. "
+            )
+
+        if offset < 0:
+            invalid_offset_error()
+
+        current_offset = 0
+        for line_index, line_offsets in enumerate(self._wrap_offsets):
+            wrapped_line_height = len(line_offsets) + 1
+            current_offset += wrapped_line_height
+            if current_offset > offset:
+                return line_index
```

Comment:
When you compute the `_wrap_offsets` you could also compute a list of the corresponding "`current_offset`" values, which would then mean this search could be done in $O(\log n)$ with the module `bisect` or your own implementation of binary search.

Reply:
I'm avoiding any caching related optimisations at this stage because they make the code brittle and hard to reason about. I do have some things in the back of my mind relating to caching, and after I pull everything together, I'll consider where it makes sense to add caching.

I suspect when wrapping is used, a document with 5 _paragraphs_ will result in a `_wrap_offsets` of length 5, and so a binary search isn't worth the burden.